### DISCUSSION TITLE
Suggestion: hidden/implicit tuple to ctor operator.

### DISCUSSION THREAD

**Discussion by Redageddon** [dotnet/csharplang#3898] [General]
OK, so every class that has a ctor with more than 1 argument should contain a hidden implicit conversion from a tupple matching every ctor. See image for example.
[![Img1](https://user-images.githubusercontent.com/53098956/93374844-2870d280-f80c-11ea-8924-fcf28514214e.png)](https://gist.github.com/rubiksmaster02/f1a8ad4c095857f8105bc0af584f7b4a)
This would allow for many different things **(this idea i got from the implicit new from c# 9, i was just thinking i could take it a step further)** such as, 
[![example1](https://user-images.githubusercontent.com/53098956/93376338-4b03eb00-f80e-11ea-8e74-7442c19f3e5f.png)](https://gist.github.com/rubiksmaster02/4c4675dde7d245570ee302fad9342ce7)
or
[![image](https://user-images.githubusercontent.com/53098956/93376515-93bba400-f80e-11ea-895b-c7d11f2d9536.png)](https://gist.github.com/rubiksmaster02/834ff3f12cc85ee16786b9f34240e868)
But wait, you might be saying, *what if i want to define my own ctor that takes in a tuple*. 
Answer: That is why the operator should be hidden.
[![image](https://user-images.githubusercontent.com/53098956/93377054-5c99c280-f80f-11ea-9792-32ca4968ff05.png)](https://gist.github.com/rubiksmaster02/d7272372585e77adb1419c46d3cb98d1)
This would be like if you had a child class override a virtual method. But, you don't actually call override on the operator or anything like that, it just would use the implemented one instead of the default hidden one. What i mean by hidden is that you dont actually see and cant edit it, the only way to change what it is, is to provide an implementation yourself. That is, yet again, the same as how a child class could use a virtual method from a parent class without actually having to override or implement it.

**Comment by HaloFour** [comment]
C# 9.0 will bring target-typed `new` expressions which I think make a lot of this proposal unnecessary:

```cs
using System;

public class TestClass {
    public TestClass(int x, string y) { }
}

public static class Program {
    public static void Main() {
        TestClass single = new(0, "zero");
        var asd = new TestClass[] {
            new(1, "one"),
            new(2, "two"),
            new(3, "three"),    
            new(4, "four"),
            new(5, "five")
        };
    }
}
```

Beyond that I think I'd rather have a way to "splat" a tuple into arguments to a constructor or method rather than something specific to implicitly convert from a tuple to a constructor of a type.

**Reply by Redageddon** [reply]
I explicitly stated that this was to go farther than c#9's feature, please read everything before making assumptions

**Reply by Redageddon** [reply]
Third sentence: (this idea i got from the implicit new from c# 9, i was just thinking i could take it a step further)

**Reply by Redageddon** [reply]
I stated that because of people like you who would bring up c#9 new, this was a feature suggestion to have you type even less than the new

**Reply by CyrusNajmabadi** [reply]
I think the point is that this doesn't provide *enough* value on top of hte value we're already getting here.  i.e. something new is being created.  It's intentional that we like that 'new' is very explicit here to call that out.  The savings then dont' seem substantial enough, and the clarity seems concerning, so i would not have this be done by default.

Note: you could always do something similar with source-generators if you wanted :)

**Reply by Redageddon** [reply]
I still think that calling new for every single item in an array creation is a waste, and that this covers more than it.

**Reply by HaloFour** [reply]
I disagree.  I think that the `new` is the most important part of that expression.  If you want implicit conversion from a tuple you can add such a conversion operator, or write a source generator to write them for you.

**Reply by marksmeltzer** [reply]
> I still think that calling new for every single item in an array creation is a waste, and that this covers more than it.

In all of the millions of lines of code I've written in my career, code initializing lists and arrays accounts for a tiny, tiny percentage of that.

The few times I would benefit from a succinct way to init arrays/lists/collections is typically in automated tests where I want the tests to use canned data. However, if I really need to init many items, I just load them from a file instead (CSV, JSON, etc.) which has a succinct syntax.
